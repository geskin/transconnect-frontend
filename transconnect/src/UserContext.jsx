import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

