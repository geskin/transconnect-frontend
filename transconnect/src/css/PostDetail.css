.post-details {
    text-align: left;
    padding-left: 20px;
}

.post-details h3,
.post-details p,
.post-details small {
    color: black;
}