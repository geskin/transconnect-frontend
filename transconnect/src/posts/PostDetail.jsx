import React, { useContext, useEffect, useState } from "react";
import UserContext from "../UserContext";
import { Link, useNavigate, useParams } from "react-router-dom";
import TransconnectApi from "../api";
import { formatDate } from "../utils/formatDate";
import CommentsBottomNavigation from "../utils/bottomNav";
import "../css/PostDetail.css";
import { Button, Typography } from "@mui/material";


/** PostDetail: shows details on a particular post including all associated comments 
 * 
 * auth required: logged in
*/

const PostDetail = () => {
    const { id } = useParams();
    console.debug(id);
    const [post, setPost] = useState({ tags: [] });
    const [user, setUser] = useState({});
    const { currUser } = useContext(UserContext);
    const navigate = useNavigate();

    if (!currUser) return <Typography>Loading...</Typography>;

    useEffect(() => {
        console.debug(currUser);

        if (!currUser) {
            navigate("/posts");
            return;
        }

        const fetchPost = async () => {
            try {
                const post = await TransconnectApi.getPost(id);
                console.debug(post);
                if (post) {
                    setPost(post);
                    setUser(post.user);
                }
            } catch (err) {
                console.error("Error fetching post", err);
            }
        };

        fetchPost();
    }, [currUser, id, navigate]);

    const handleDeletePost = async () => {
        try {
            await TransconnectApi.deletePost(id);
            navigate("/posts");
        } catch (err) {
            console.error("Error deleting post", err);
        }
    };

    return (
        <div className="col-md-8 offset-md-2">
            <div className="post-details">
                <h4><Link to={`/users/${user.username}`}>@{user.username}</Link></h4>
                <h3 className="mb-3">{post.title}</h3>
                <p>{post.content}</p>
                {post.tags && post.tags.length > 0 && (
                    <div>
                        {post.tags.map(t => (
                            <Button variant="outlined" disabled key={t.name}>
                                {t.name}
                            </Button>
                        ))}
                    </div>
                )}
                <small><i>Posted {formatDate(post.createdAt)}</i></small>
                {(currUser.role === "ADMIN" || currUser.username === user.username) && (
                    <div>
                        <Button
                            variant="contained"
                            color="error"
                            onClick={handleDeletePost}
                            size="small"
                        >
                            Delete
                        </Button>
                        <Button onClick={() => navigate(`/posts/${id}/edit`)} size="small">
                            Edit
                        </Button>
                    </div>
                )}

            </div>
            <hr></hr>
            <div>
                <CommentsBottomNavigation postId={id} />
            </div>
        </div>
    );
}

export default PostDetail;
