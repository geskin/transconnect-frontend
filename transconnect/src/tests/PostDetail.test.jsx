import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import PostDetail from "../posts/PostDetail";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import CommentsBottomNavigation from "../utils/bottomNav";

vi.mock("../api");
vi.mock("../utils/bottomNav", () => ({
    __esModule: true,
    default: vi.fn(() => <div data-testid="comments-nav">CommentsNavigation</div>),
}));

describe("PostDetail", () => {
    const mockCurrUser = { username: "testuser", id: 1, role: "USER" };
    const mockPost = {
        id: 101,
        title: "Test Post",
        createdAt: "2024-03-14T12:00:00Z",
        content: "This is a test post content.",
        tags: [{ name: "React" }, { name: "Vitest" }],
        user: { username: "otheruser" },
    };

    beforeEach(() => {
        TransconnectApi.getPost.mockResolvedValue(mockPost);
    });

    it("renders without crashing", () => {
        render(
            <MemoryRouter initialEntries={["/posts/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/posts/:id" element={<PostDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter initialEntries={["/posts/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/posts/:id" element={<PostDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("fetches and displays post details", async () => {
        render(
            <MemoryRouter initialEntries={["/posts/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/posts/:id" element={<PostDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText("Test Post")).toBeInTheDocument();
            expect(screen.getByText("This is a test post content.")).toBeInTheDocument();
        });

        expect(screen.getByTestId("comments-nav")).toBeInTheDocument();
    });

    it("displays tags correctly", async () => {
        render(
            <MemoryRouter initialEntries={["/posts/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/posts/:id" element={<PostDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText("React")).toBeInTheDocument();
            expect(screen.getByText("Vitest")).toBeInTheDocument();
        });
    });

    it("handles post deletion", async () => {
        TransconnectApi.deletePost.mockResolvedValue({});

        render(
            <MemoryRouter initialEntries={["/posts/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/posts/:id" element={<PostDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText("Test Post")).toBeInTheDocument();
        });

        fireEvent.click(screen.getByText("Delete"));

        await waitFor(() => {
            expect(TransconnectApi.deletePost).toHaveBeenCalledWith("101");
        });
    });
});
