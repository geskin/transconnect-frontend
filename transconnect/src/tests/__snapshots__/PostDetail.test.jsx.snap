// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostDetail > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-md-8 offset-md-2"
  >
    <div
      class="post-details"
    >
      <h4>
        <a
          data-discover="true"
          href="/users/undefined"
        >
          @
        </a>
      </h4>
      <h3
        class="mb-3"
      />
      <p />
      <small>
        <i>
          Posted Invalid date
        </i>
      </small>
    </div>
    <hr />
    <div>
      <div
        data-testid="comments-nav"
      >
        CommentsNavigation
      </div>
    </div>
  </div>
</DocumentFragment>
`;
